Source/WebCore/rendering/RenderBlockFlow.cpp

```diff
@@ -1367,6 +1394,55 @@ bool RenderBlockFlow::shouldTrimChildMargin(MarginTrimType marginTrimType, const
     return lastInFlowChildBox() == &child;
 }
 
+void RenderBlockFlow::trimFloatBlockEndMargins(LayoutUnit blockFormattingContextInFlowContentHeight)
+{
+    ASSERT(m_floatingObjects && establishesBlockFormattingContext());
+    auto computeFloatMarginAfterStartLocationInBlockFormattingContext = [&](const RenderBox& floatBox) {
+        // Start with the location of the float within its containing block. If this containing
+        // block is the BFC in which we are trimming, then we can use this as the starting location
+        // since this is the float's positin within the BFC. If not, add the containing block's location
+        // to the float's location. This new value should be the location of the float relative to this
+        // new containing block. Keep doing this until we find the containing block that establishes
+        // a BFC
+        auto bfcRoot = floatBox.blockFormattingContextRoot();
+        ASSERT(bfcRoot);
+
+        auto startLocation = floatBox.localToAbsolute();
+        auto containingBlock = floatBox.containingBlock();
+        while (!containingBlock->establishesBlockFormattingContext()) {
+            startLocation += containingBlock->localToAbsolute();
+            containingBlock = containingBlock->containingBlock();
+        }
+        if (bfcRoot->isHorizontalWritingMode())
+            return LayoutUnit { startLocation.y() + bfcRoot->logicalHeightForChild(floatBox) };
+        return LayoutUnit { startLocation.x() + bfcRoot->logicalWidthForChild(floatBox) };
+    };
+    for (auto& floatingObject : m_floatingObjects->set()) {
+        auto& floatBox = floatingObject->renderer();
+        if (!floatBox.containingBlock()->style().marginTrim().contains(MarginTrimType::BlockEnd) || !floatBox.marginAfter())
+            continue;
+
+        // There are 3 different cases that need to be taken into consideration when trimming the
+        // block-end margin of a float:
+        // 1.) The float margin box extends past the deepest past of content (margin should be trimmed all the way)
+        // 2.) The end of the deepest piece of is in between the float's border box and marign box (marign should be trimmed up to the content)
+        // 3.) The deepest piece of content is below the float's margin box (no trimming occurs)
+        auto floatMarginAfterStartLocation = computeFloatMarginAfterStartLocationInBlockFormattingContext(floatBox);
+        if (floatMarginAfterStartLocation >= blockFormattingContextInFlowContentHeight)
+            trimMarginForFloat(*floatingObject, MarginTrimType::BlockEnd, floatBox.marginAfter(&style()));
+        else if (blockFormattingContextInFlowContentHeight > floatMarginAfterStartLocation && (blockFormattingContextInFlowContentHeight < floatMarginAfterStartLocation + floatBox.marginAfter()))
+            trimMarginForFloat(*floatingObject, MarginTrimType::BlockEnd, floatMarginAfterStartLocation + floatBox.marginAfter(&style()) - blockFormattingContextInFlowContentHeight);
```

Comment:
I don't think you need to pass in the box's own style to marginAfter.

Reply:
Just to make sure, does box here refer to the containing block? That's what I was trying to do so that the correct margin was retrieved using the writing mode of the containing block

Reply:
ok, it was super not obvious (this is a terrible interface). Also, you don't seem to be calling it with the containing block's style but with the BFC root's style (which in practice when it comes to figuring out the writing-mode does not make any difference, but as you can see I made a mistake as the reviewer and I think you did as the author, so we should probably refrain from using it and instead do something more explicit).

Reply:
Oh wow good catch. I went ahead and stored the float's containing block into its own variable since we use it multiple times here. With that I can use the `marginAfterForChild` on `RenderBlock` which should get the appropriate value. It's a little more explicit since we are using the containing block directly.